Combining totems in a crafting grid creates a new random totem.

CplPibald opened the pull request:
Fixes #37

Adds a custom recipe that takes N totems (configurable) in a crafting grid and produces
a new totem with a random level 1 ability.

Open issues:
1) No JEI integration.  Recipe does not show up in JEI
2) Crafting output slot shows a totem with a random ability that is usually not the random ability as you get when you take the item.  A better solution would show a totem with ability text "???" when in the crafting output slot, but `IRecipe.getCraftingResult()` doesn't distinguish between the ItemStack in the output slot and in the picked up result.

CplPibald committed:
Combining totems in a crafting grid creates a new random totem.
Fixes #37

Adds a custom recipe that takes N totems (configurable) in a crafting grid and produces
a new totem with a random level 1 ability.

rubensworks commented on src/main/java/org/cyclops/everlastingabilities/EverlastingAbilities.java:
```diff
@@ -203,6 +205,10 @@ public ICapabilityProvider createProvider(IAnimals host) {
             }
         });
+		
+		if (ItemAbilityTotemConfig.totemCraftingCount > 0) {
```
Use spaces instead of tabs.

rubensworks commented on src/main/java/org/cyclops/everlastingabilities/item/ItemAbilityTotem.java:
```diff
@@ -62,6 +63,12 @@ public ItemStack getTotem(Ability ability) {
         return itemStack;
     }
 
+	public static ItemStack getRandomTotem(java.util.Random rand) {
```
Move this to `AbilityHelpers`.

Also, instead of choosing a random `EnumRarity`, perhaps it would make more sense that the rarities of the inserted totems are taken into account for choosing the new rarity. By for example taking the max/min rarity, with a small chance of increasing the rarity. This would make this feature a bit more balanced.

rubensworks reviewed (changes requested):
Looks good, I just have two comments.

rubensworks commented:
Regarding the JEI integration, this is no problem.

For the '???' text, you could do this by checking `world.isRemote` to see if you're running client-side. If the crafting result is retrieved client-side, this means that the result is just previewed, which means that you should then show `???` as ability name (or those shifting Minecraft symbols would even be nicer).

CplPibald commented:
I have all of the suggestions implemented except for the obfuscated text on the totem, which is proving quite difficult.  The code was never meant to display ability text which is not meant to be added to a player.  All of the solutions I've considered involve touching or creating a lot of code, so I'd like your suggestion how to proceed.

Originally I tried to create a dummy AbilityType with an obfuscated display name and add it to a totem.  This seemed like the cleanest solution, but unfortunately, it didn't display because the dummy type wasn't in the `AbilityTypes.REGISTRY` collection.  To solve this, I can register the dummy type as an AbilityType, but that introduces a risk of a bug causing it being used as an ability on items or players.  To protect against this, I could add a property `displayOnly` to AbilityType which is checked when trying to add the ability to an entity.

Alternatively, removing the registry check in `AbilityStoreStorage.readNBT()` solves this issue.  This would be my preferred solution to minimize impact to the rest of the code, but it also makes the code less resistant to corrupt NBT tags.  

Finally I've also looked into abandoning the dummy AbilityType idea and instead creating a new Item with the same model and texture as the totem and a custom `addInformation()`.  That seems like a lot of code for one line of text, but may be the least invasive option.

Do you have a recommendation how to do this?

rubensworks commented:
What about adding a `isObfuscated()` method to `IAbilityStore`, instead of having to do this on ability-level? A `ItemAbilityTotem#getObfuscatedTotem` can then be added to create a mutable ability store where this obfuscation field is set to `true`. A check for this can be added in `ItemGuiAbilityContainer#addInformation()`.

CplPibald committed:
Update for totem combining recipe. (issue #37)
Breaking API change: Add DisplayType property to IAbilityStore.
  Add obfuscated display type.
  Show obfuscated totem abilities until player takes totem out of crafting grid.
Move getRandomTotem to AbilityHelpers
Output rarity is now determined by inputs.  The rarity is taken from one input chosen at random.
Convert tabs to spaces

CplPibald committed:
Add percent chance (configurable) of bumping rarity by one when combining totems to get a new one.
Change tabs to spaces again

CplPibald commented:
I updated the PR with all of the recommended changes, including obfuscated totem text in the crafting window, and rarity of output based on rarity of inputs.

This now includes a breaking API change.  One new method `getDisplayType()` added to IAbilityStore, and an additional method `setDisplayType()` added to IMutableAbilityStore.  It takes an enum DisplayType, which currently only has values `NORMAL` and `OBFUSCATED`.

As for rarity of output item, my solution that didn't require a lot of enum math was to select at random one of the inputs and use its rarity, then add one 15% of the time (percentage is configurable)  The best chance to get a high-rarity output is still to use high-rarity inputs.

rubensworks commented on src/main/java/org/cyclops/everlastingabilities/item/ItemGuiAbilityContainer.java:
```diff
@@ -53,15 +53,30 @@ protected ItemGuiAbilityContainer(ExtendedConfig eConfig) {
-        if (empty) {
-            list.add(TextFormatting.GRAY.toString() + TextFormatting.ITALIC + L10NHelpers.localize("general.everlastingabilities.empty"));
+
+        switch(abilityStore.getDisplayType()) {
```
I don't really like switch cases.
What about delegating to an `addInformation` method to DisplayType and moving this logic in there?
This will also make it safer in the future when other display types would be added, that they automatically will require `addInformation` to be implemented.

rubensworks commented on src/main/java/org/cyclops/everlastingabilities/recipe/TotemRecycleRecipe.java:
```diff
@@ -0,0 +1,108 @@
+    
+    @Override
+    public ItemStack getCraftingResult(InventoryCrafting invCrafting) {
+        if (_world.isRemote) {
```
CyclopsCore has a method `MinecraftHelpers.isClientSide()`, this would avoid having to store the world.
Storing the world has two downsides:

1. If an external mod calls `getCraftingResult` without first calling `matches`, the world might still be `null`.
2. This will cause worlds to leak if they are unloaded but still stored this recipe.

CplPibald replied:
I agree, and wasn't entirely happy with caching the world.  Thanks for letting me know about isClientSide()

The second reason for storing the world was to have access to a Random object.  I don't think it is dangerous to create a new one just for one method, but it doesn't seem like a clean solution.  Does CyclopsCore have a getRandom() helper?

rubensworks replied:
I guess it's fine to just create a private (possibly static) Random instance inside the recipe, and just use that one.

rubensworks reviewed (changes requested):
This looks really good!
Thanks for the changes, I just have two more minor comments.

CplPibald committed:
Address feedback for #38
Rename AbilityStore.DisplayType enum to AbilityStoreDisplayType class
Move addInformation functionality from ItemGuiAbilityContainer to AbilityStoreDisplayType
Don't cache World object in TotemRecycleRecipe

CplPibald commented:
One more try.

Renamed AbilityStore.DisplayType enum to AbilityStoreDisplayType class
Factored out addInformation into AbilityStoreDisplayType
Removed caching of World object in TotemRecycleRecipe

rubensworks commented:
Perfect, thanks!

Can you send me your curseforge username? I'll add you as contributor to the curseforge project.

CplPibald commented:
My curseforge name is also CplPibald.

... and I just realized I left a bug in `TotemRecycleRecipe.java`, line 99.  It returns constant 3 when it should return `ItemAbilityTotemConfig.totemCraftingCount`.  Oops :-(  I don't know what getRecipeSize() is used for; only that it's added by Forge.

rubensworks commented:
@CplPibald Added you to the project :-)

Oh oops, can you make an issue or PR for that?
It's not something urgent I guess.
